### DISCUSSION TITLE
magit-status-quick: accept optional directory?

### DISCUSSION THREAD

**Discussion by isker** [magit/magit#4636] [Ideas]
`magit-status` accepts an optional directory: https://github.com/magit/magit/blob/3525928602a3b68855fa239cadbc7d31c67fd50e/lisp/magit-status.el#L236

`magit-status-quick` does not: https://github.com/magit/magit/blob/3525928602a3b68855fa239cadbc7d31c67fd50e/lisp/magit-status.el#L313

As in the original issue #4259, I'm interested in using `magit-status-quick` as a hook for switching projects.  In Doom Emacs specifically, this is done with a function that takes an argument:
https://github.com/hlissner/doom-emacs/blob/42e5763782fdc1aabb9f2624d468248d6978abe2/modules/ui/workspaces/config.el#L13-L16

`magit-status` is able to be used to this effect (and is, [in a different place](https://github.com/hlissner/doom-emacs/blob/42e5763782fdc1aabb9f2624d468248d6978abe2/modules/completion/vertico/autoload/magit.el#L5-L8)), but not `magit-status-quick`.

I don't have a good command of elisp, so I'm wondering if there's any "obvious" reason why `magit-status-quick` can't do this.  If not, would you be open to this enhancement?

**Comment by isker** [comment]
Re: my poor command of elisp, one thing I don't understand about `magit-status-quick` is this: https://github.com/magit/magit/blob/3525928602a3b68855fa239cadbc7d31c67fd50e/lisp/magit-status.el#L329 Does `interactive` magically define `buffer`, or something else?

**Reply by isker** [reply]
I misread this.  It’s if-let 😁.  I’ll have to read more code tomorrow to decide whether this branch could handle a directory argument…

**Comment by tarsius** [comment]
A quick draft, untested and unfinished:

```elisp
(defun magit-project-status-quick (directory)
  (let ((toplevel (magit-toplevel directory)))
    (setq directory (file-name-as-directory
                     (expand-file-name directory)))
    (if (and toplevel (file-equal-p directory toplevel))
        (let ((default-directory toplevel))
          (if-let ((buffer (and (not (magit-get-mode-buffer 'magit-status-mode
                                                            nil 'selected))
                                (magit-get-mode-buffer 'magit-status-mode))))
              (magit-display-buffer buffer)
            (magit-status-setup-buffer)))
      ;; TODO Look at the various project switchers and decide what to do here.
      )))
```

**Reply by isker** [reply]
Seems to work when set as Doom's `+workspaces-switch-project-function`.

**Reply by isker** [reply]
I'm not sure this actually works when switching from another magit repository - it seems that `magit-get-mode-buffer` would have to be able to be parameterized with the directory as well:

https://github.com/magit/magit/blob/613682a70631178caee15d1b5a16fae90da4af7b/lisp/magit-mode.el#L838-L839

Right now it cannot possibly find the other repository's buffers when trying the fast case.

**Comment by isker** [comment]
I played around with this diff:
```diff
diff --git a/lisp/magit-mode.el b/lisp/magit-mode.el
index 95d4898..f08635d 100644
--- a/lisp/magit-mode.el
+++ b/lisp/magit-mode.el
@@ -831,7 +831,7 @@ into thinking a buffer belongs to a repo that it doesn't.")
 (defvar-local magit-buffer-locked-p nil)
 (put 'magit-buffer-locked-p 'permanent-local t)
 
-(defun magit-get-mode-buffer (mode &optional value frame)
+(defun magit-get-mode-buffer (mode &optional value frame directory)
   "Return buffer belonging to the current repository whose major-mode is MODE.
 
 If no such buffer exists then return nil.  Multiple buffers with
@@ -848,7 +848,7 @@ If `visible', then only consider buffers on all visible frames.
 If `selected' or t, then only consider buffers on the selected
   frame.
 If a frame, then only consider buffers on that frame."
-  (if-let ((topdir (magit-toplevel)))
+  (if-let ((topdir (magit-toplevel directory)))
       (cl-flet* ((b (buffer)
                     (with-current-buffer buffer
                       (and (eq major-mode mode)
diff --git a/lisp/magit-status.el b/lisp/magit-status.el
index eed0dcd..2e37718 100644
--- a/lisp/magit-status.el
+++ b/lisp/magit-status.el
@@ -310,7 +310,7 @@ also contains other useful hints.")
 (put 'magit-status-here 'interactive-only 'magit-status-setup-buffer)
 
 ;;;###autoload
-(defun magit-status-quick ()
+(defun magit-status-quick (&optional directory)
   "Show the status of the current Git repository, maybe without refreshing.
 
 If the status buffer of the current Git repository exists but
@@ -328,10 +328,10 @@ init file: (global-set-key (kbd \"C-x g\") 'magit-status-quick)."
   (interactive)
   (if-let ((buffer
             (and (not current-prefix-arg)
-                 (not (magit-get-mode-buffer 'magit-status-mode nil 'selected))
-                 (magit-get-mode-buffer 'magit-status-mode))))
+                 (not (magit-get-mode-buffer 'magit-status-mode nil 'selected directory))
+                 (magit-get-mode-buffer 'magit-status-mode nil nil directory))))
       (magit-display-buffer buffer)
-    (call-interactively #'magit-status)))
+    (call-interactively #'magit-status))) ;; TODO what to do here?
 
 (defvar magit--remotes-using-recent-git nil)
```

But there are a few problems.  The first is that TODO; the interactive/noninteractive split of this function is getting ugly.

The bigger problem is more specific to Doom:  it seems that Doom invokes your `+workspaces-switch-project-function` _after_ having switched to the [perspective](https://github.com/Bad-ptr/persp-mode.el) of the target workspace.  If the top buffer of that perspective was the magit-status-mode buffer for the project, then it will trigger the "status buffer already selected, actually refresh" logic of `magit-status-quick`, thus defeating the whole purpose.

I think the easiest solution for me is to have a distinct "never refresh" version of `magit-status-quick`; the behavior of `magit-status-quick` existing function is really designed to satisfy an interactive command:

```emacs-lisp
(defun magit-status-quicker (&optional directory)
  (if-let ((buffer (magit-get-mode-buffer 'magit-status-mode nil nil directory)))
      (magit-display-buffer buffer)
    (magit-status-setup-buffer directory)))
```

But it still does need that patch to `magit-get-mode-buffer` above.

This is a bit of a mess, I doubt a new function is ideal to be contributed back to magit.

**Reply by tarsius** [reply]
> the interactive/noninteractive split of this function is getting ugly.

I intend to keep it. As we see here different project switchers and different starter-kits that use those switchers may wish to call the "switch function" in different, mutually exclusive, ways and trying to cope with that on Magit's side by providing one function that covers all those different expectations is not feasible.

I am working on other things right now, but I will get to this too eventually.

**Comment by tarsius** [comment]
I've add an optional DIRECTORY argument to `magit-status-quick`.

**Reply by isker** [reply]
Thank you!

**Reply by tarsius** [reply]
Sorry that it took a while. ;P

**Reply by isker** [reply]
I don’t remember this at all!
